java/memory/src/main/java/org/apache/arrow/memory/SegmentAllocator.java

```diff
@@ -0,0 +1,146 @@
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.arrow.memory;
+
+import org.apache.arrow.util.Preconditions;
+
+import io.netty.buffer.ArrowBuf;
+import io.netty.util.internal.OutOfDirectMemoryError;
+
+/**
+ * A segment allocator allocates a buffer whose size is a multiple of the segment size.
+ */
+public class SegmentAllocator extends BaseAllocator {
+
+  protected final int segmentSize;
+
+  public SegmentAllocator(final long limit, final int segmentSize) {
+    this(AllocationListener.NOOP, limit, segmentSize);
+  }
+
+  public SegmentAllocator(final AllocationListener listener, final long limit, final int segmentSize) {
+    this(null, listener, "Segment Allocator", 0, limit, segmentSize);
+  }
+
+  /**
+   * Initialize an allocator
+   * @param parentAllocator   parent allocator. null if defining a root allocator
+   * @param listener          listener callback. Must be non-null -- use
+   *                          {@link AllocationListener#NOOP} if no listener desired
+   * @param name              name of this allocator
+   * @param initReservation   initial reservation. Cannot be modified after construction
+   * @param maxAllocation     limit. Allocations past the limit fail. Can be modified after construction
+   * @param segmentSize segment size. Must be greater than or equal to 1024, and must be a power of 2.
+   */
+  public SegmentAllocator(
+          final BaseAllocator parentAllocator,
+          final AllocationListener listener,
+          final String name,
+          final long initReservation,
+          final long maxAllocation,
+          final int segmentSize) {
+    super(parentAllocator, listener, name, initReservation, maxAllocation);
+
+    if (segmentSize < 1024L) {
+      throw new IllegalArgumentException("The segment size cannot be smaller than 1024");
+    }
+    if ((segmentSize & (segmentSize - 1)) != 0) {
+      throw new IllegalArgumentException("The segment size must be a power of 2");
+    }
+    this.segmentSize = segmentSize;
+  }
+
+  @Override
+  public ArrowBuf buffer(final int initialRequestSize, BufferManager manager) {
+    assertOpen();
+
+    Preconditions.checkArgument(initialRequestSize >= 0, "the requested size must be non-negative");
+
+    if (initialRequestSize == 0) {
+      return empty;
+    }
+
+    // round to next largest multiple of the segment size
+    final int actualRequestSize =
```

Comment:
It seems like this is the only thing that changes between this and the method you are overriding.  Can you factor out that code and this into a rounding policy?  Then expose a new constructor on BaseAllocator that takes the policy.  Then the new allocator would propagate the policy?  I think that would reduce the size of this PR dramatically.

Reply:
This is really good suggestion. Many thanks!

Maybe you have also noticed that, there is some overlap with [PR-4358](https://github.com/apache/arrow/pull/4358). Do you think I should also revise that PR accordingly?

Reply:
@liyafan82 I think #4358 is different because it changes the public API of consumers of the class.  In this case you are only adding a constructor.  I'm more comfortable with this change as long as there aren't performance regressions, and I need to confirm there isn't something subtle about the power of 2 rounding policy?

CC @pravindra @praveenbingo  is there a strong reason all rounding was power of 2 or was it just convenience?

Reply:
@liyafan82 sorry to be clear, I don't think you should revise the other PR until we come to consensus on the mailing list that it should be added (I don't want to cause any wasted work, and I appreciate all the effort you have put in being responsive to comments).

Reply:
> @liyafan82 I think #4358 is different because it changes the public API of consumers of the class. In this case you are only adding a constructor. I'm more comfortable with this change as long as there aren't performance regressions, and I need to confirm there isn't something subtle about the power of 2 rounding policy?
> 
> CC @pravindra @praveenbingo is there a strong reason all rounding was power of 2 or was it just convenience?

I am also more comfortable with this change, thanks to your suggestion. 
I think power-of-2 rounding have two benefits:
1. Better performance - whenever a part of a big segment is allocated, the remaining part is still the sum of some valid segments.
2. Reduce memory fragmentation - this leads to more efficient memory usage.

Reply:
> @liyafan82 sorry to be clear, I don't think you should revise the other PR until we come to consensus on the mailing list that it should be added (I don't want to cause any wasted work, and I appreciate all the effort you have put in being responsive to comments).

Thanks for your consideration. Let's wait for the feedback from the mailing list.